rdkPlugins/Networking/source/PortForwarding.cpp

```diff
@@ -0,0 +1,693 @@
+                 ipAddress.c_str(), "32",
+                 protocol.c_str(),
+                 protocol.c_str(),
+                 portNumber.c_str(),
+                 id.c_str());
+    }
+    else if (ipVersion == AF_INET6)
+    {
+        snprintf(buf, sizeof(buf), baseRule.c_str(),
+                 ipAddress.c_str(), "128",
+                 protocol.c_str(),
+                 protocol.c_str(),
+                 portNumber.c_str(),
+                 id.c_str());
+    }
+    else
+    {
+        return std::string();
+    }
+
+    return buf;
+}
+
+
+// -----------------------------------------------------------------------------
+/**
+ *  @brief Construct rules to port forward from container to host and add them
+ *  to the 'ruleSets' vector.
+ *
+ *  NB: these rules should be inserted to the end of their respective tables,
+ *  so we put the new rules into the ruleSets vector's index 0 reserved for
+ *  insert(-I) rules.
+ *
+ *  @param[in]  ruleSets            Vector to place new rulesets in.
+ *  @param[in]  containerId         Container identifier.
+ *  @param[in]  containerAddress    IP address of the container.
+ *  @param[in]  vethName            The container's assigned veth device name.
+ *  @param[in]  ports               Ports to forward.
+ *  @param[in]  len                 Number of ports/protocol combinations.
+ *  @param[in]  ipVersion           IPv family version (AF_INET/AF_INET6).
+ *
+ *  @return returns true on success, otherwise false.
+ */
+bool constructContainerToHostRules(std::vector<Netfilter::RuleSet> &ruleSets,
+                                   const std::string &containerId,
+                                   const std::string &containerAddress,
+                                   const std::string &vethName,
+                                   rt_defs_plugins_networking_data_port_forwarding_container_to_host_element **ports,
+                                   size_t len, const int ipVersion)
+{
+    std::list<std::string> natRules;
+    std::list<std::string> filterRules;
+
+    // construct rules for each port to enable forwarding on
+    for (int i = 0; i < len; i++)
+    {
+        std::string port = std::to_string(ports[i]->port);
+
+        // default to tcp if no protocol is set
+        std::string protocol;
+        if (ports[i]->protocol)
+        {
+            protocol = ports[i]->protocol;
+
+            // transform to lower case to allow both upper and lower case entries
+            std::transform(protocol.begin(), protocol.end(), protocol.begin(), ::tolower);
+
+            // check for accepted protocol values
+            if (strcmp(protocol.c_str(), "tcp") != 0 && strcmp(protocol.c_str(), "udp") != 0)
+            {
+                AI_LOG_ERROR("invalid protocol value '%s' for port at index %d",
+                             ports[i]->protocol, i);
+                return false;
+            }
+        }
+        else
+        {
+            protocol = "tcp";
+        }
```

Comment:
This block is duplicated from `constructHostToContainerRules`.

Perhaps refactor into `getProtocol` method?